### ISSUE TITLE
disassembler enhancement request

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#147]
When we disassmble mips & mips64 relative branch instructions, they look like this:

```
0x4eb4f63c   124  10200007       beq     at, zero_reg, 7
0x4eb4f640   128  00000000       nop
0x4eb4f644   132  3c190000       lui     t9, 0x0             ;; code: BUILTIN
0x4eb4f648   136  37394eb4       ori     t9, t9, 0x4eb4
....
```

The 7 is extracted from the instruction encoding, but for the human to figure out where it is going, they have to run the calculation of (pc_ + 4) + (branch_distance << 2). Annoying to calculate all the time. For larger distances you can't just count instructions.

The other arch's do a much more human friendly job. They give absolute address of destination, plus usually also the pc-realtive offset from start of function, which is very easy to parse by eye. Here is arm32:

```
                  -- B3 start (deconstruct frame) --
0x5b83cce0    96  e59a0014       ldr r0, [r10, #+20]
0x5b83cce4   100  e1a0d00b       mov sp, fp
0x5b83cce8   104  e8bd4800       ldmia sp!, {fp, lr}
0x5b83ccec   108  e28dd004       add sp, sp, #4
0x5b83ccf0   112  e12fff1e       bx lr
                  -- B4 start (no frame) --
                  -- B1 start (deferred) --
0x5b83ccf4   116  e3071c0c       movw r1, #31756
0x5b83ccf8   120  e3401b50       movt r1, #2896
0x5b83ccfc   124  e3a00000       mov r0, #0
0x5b83cd00   128  e51b7010       ldr r7, [fp, #-16]
0x5b83cd04   132  e59fc01c       ldr ip, [pc, #+28]          ;; code: STUB, CEntryStub, minor: 0
0x5b83cd08   136  e12fff3c       blx ip
0x5b83cd0c   140  eafffff3       b -44 -> 96  (0x5b83cce0)
```

You get to that last instruction and see immediately where it is going. You don't even have to look at the 32-bit hex address. x64 does something very similar.

Please update the disassembler to do this.

**Comment by paul99** [comment]
I can't give this a pri-1 in good conscience, however, this should be a short-duration task, and just take a few hours. Let's get it done. It will make debugging our harder problems a lot easier.

**Comment by ilijapavlovic** [comment]
Disassembler enhancement implemented.
https://codereview.chromium.org/1213553004/

**Comment by paul99** [comment]
Code review for above: output format is exactly what I wanted but the branch offset is off by one instruction.

From manual: "An 18-bit signed offset (the 16-bit offset field shifted left 2 bits) is added to the address of the instruction following the branch (not the branch itself), in the branch delay slot, to form a PC-relative effective target address."

Here is some disassembled from `d8` startup with --print-all-code:

```
--- End code ---
kind = HANDLER
major_key = <NoCache>Stub
name = Debug
compiler = unknown
Instructions (size = 108)
0x4dd23380     0  27bdfff8            addiu   sp, sp, -8
0x4dd23384     4  afa70004            sw      a3, 4(sp)
0x4dd23388     8  afa40000            sw      a0, 0(sp)
0x4dd2338c    12  3c0439c6            lui     a0, 0x39c6          ;; object: 0x39c6307d Cell for 0
0x4dd23390    16  3484307d            ori     a0, a0, 0x307d
0x4dd23394    20  8c840003            lw      a0, 3(a0)
0x4dd23398    24  1480000d            bne     a0, zero_reg, 13 -> 76  (0x4dd233cc)
0x4dd2339c    28  00000000            nop
0x4dd233a0    32  3c0439c6            lui     a0, 0x39c6          ;; object: 0x39c63085 WeakCell for 0x39c62769 PropertyCell .....
0x4dd233a4    36  34843085            ori     a0, a0, 0x3085
0x4dd233a8    40  8c840003            lw      a0, 3(a0)
0x4dd233ac    44  30810001            andi    at, a0, 0x1
0x4dd233b0    48  10200007        --  beq     at, zero_reg, 7 -> 76  (0x4dd233cc)
0x4dd233b4    52  00000000       |    nop
0x4dd233b8    56  8c840007       |    lw      a0, 7(a0)
0x4dd233bc    60  27bd0008       |    addiu   sp, sp, 8
0x4dd233c0    64  03e00008       |    jr      ra
0x4dd233c4    68  00801025       |    mov     v0, a0
0x4dd233c8    72  10000008       |    beq     zero_reg, zero_reg, 8 -> 104  (0x4dd233e8)
0x4dd233cc    76  00000000        --> nop
0x4dd233d0    80  8fa40000            lw      a0, 0(sp)
0x4dd233d4    84  8fa70004            lw      a3, 4(sp)
0x4dd233d8    88  27bd0008            addiu   sp, sp, 8
0x4dd233dc    92  3c192822            lui     t9, 0x2822          ;; code: BUILTIN
0x4dd233e0    96  3739e660            ori     t9, t9, 0xe660
0x4dd233e4   100  03200008            jr      t9
0x4dd233e8   104  00000000            nop
```

You show the `beq` branching to offset 76, but it should branch to offset 80, 7 instructions after the `nop` in the branch delay slot of the `beq`

Other examples in this function ....

**Comment by ilijapavlovic** [comment]
Corrections are done on following instructions:
- bc1t, bc1f, bltz, bltzal, bgez, bgezal, bgezall, beq, bne, blez, bgtz.

From the specification, the common thing for all of them is following:

```
I+1:         PC <- PC + target_offset
```

**Comment by ilijapavlovic** [comment]
Landed.
https://codereview.chromium.org/1213553004/

**Comment by paul99** [comment]
Re-opening. This CL causes a bug in mips32 cctest/test-disasm-mips/Type0, debug on sim or hardware.

https://v8.mips.com/bbv8/builders/Sim%20-%20Debug%20-%20edge/builds/7533

```
# Fatal error in ../src/mips/assembler-mips.cc, line 1411
# Check failed: in_range && ((target & 3) == 0).
```

**Comment by ilijapavlovic** [comment]
###### MIPS32

In functions j(int32_t target) and jal(int32_t target) there is the same segment of code:

```
#ifdef DEBUG
  // Get pc of delay slot.
  uint32_t ipc = reinterpret_cast<uint32_t>(pc_ + 1 * kInstrSize);
  bool in_range = (ipc ^ static_cast<uint32_t>(target) >>
                  (kImm26Bits + kImmFieldShift)) == 0;
  DCHECK(in_range && ((target & 3) == 0));
#endif
```

The problem is in following condition:

```
(ipc ^ static_cast<uint32_t>(target) >> (kImm26Bits + kImmFieldShift)) == 0
```

After replacing constants, we have following simplified expression:

```
(ipc ^ target >> 28) == 0
```

This expression checks the target range. 
Important thing here is that operator `>>` has higher priority than operator `^`.

Consider following (simplified) expressions:

```
res1 = target >> 28
res2 = ipc ^ res1
```

and condition `res2 == 0`.

If the target is in range:

```
res1 = 0
res2 = ipc ^ res1 = ipc ^ 0 = ipc
```

If the target is out of range:

```
res1 != 0
res2 = ipc ^ res1 = <altered the least significant bits in ipc>
```

Therefore, conclusion is that following condition is wrong

```
(ipc ^ static_cast<uint32_t>(target) >> (kImm26Bits + kImmFieldShift)) == 0
```

and it should be

```
(ipc ^ static_cast<uint32_t>(target) >> (kImm26Bits + kImmFieldShift)) == ipc
```

###### MIPS64

There is little bit different expression/condition:

```
    bool in_range = ((ipc ^ static_cast<uint64_t>(target)) >>
                     (kImm26Bits + kImmFieldShift)) == 0;
```

or simplified:

```
in_range = ((ipc ^ target) >> 28) == 0;
```

If ipc has a value which can be stored in 28-bits, this condition will be OK,
but in case where the ipc value needs more than 28-bits and the target is 28-bits,
then this condition is wrong (or is not well defined).

Well defined condition should/could be:

```
in_range = (ipc ^ target >> 28) == ipc;
```

**Comment by paul99** [comment]
OK, I definitely see the operator precedence problem -- I'm kind of amazed that worked for so long without failure.

To me, the mips64 case looks exactly right (simplified):

```
in_range = ((ipc ^ target) >> 28) == 0;
```

You're trying to test are the upper 4 bits (d31:d28) _different_ on ipc and target. The expression does that.

The expression you proposed: `in_range = (ipc ^ target >> 28) == ipc;` does _not_ do that. It is shifting only target, and doing an xor with unshifted address, I don't see how that can work.

I think we just need to add the proper parenthesis to fix the precedence issue.

**Comment by ilijapavlovic** [comment]
I suppose that it worked because of small target values which are used in 
test suites. In disassembler test we use a maximal value for the target and 
such case was not tested till now.

Here are two cases which will show what is wrong with condition
`((ipc ^ target) >> 28) == 0`

Case 1:

```
ipc    = 0x10000000
target = 0x0ffffffc     // the target has max value

res      = ipc ^ target = 0x1ffffffc
in_range = res >> 28 = 0x1        // != 0 => Error; the target is in range!
```

Case 2:

```
ipc    = 0x10000000
target = 0x1ffffffc     // the target is out of range

res      = ipc ^ target = 0x0ffffffc
in_range = res >> 28 = 0x0        // == 0 => Error; the target is out of range!
```

The condition `(ipc ^ target >> 28) == ipc` will do evaluation on right way for values defined in previous cases.

Actually, it is not clear to me why the ipc is used at all for checking target range?
For me, the condition  `(target >> 28) == 0` would do expected checking and 
is clear enough.

**Comment by paul99** [comment]
Please take another look at the instruction set reference for `j` and `jal`. These are _very_ weird instructions. They are _not_ pc-relative instruction. The `target` value  that is passed is a 32-bit pointer of the address to jump to. [1]

However, the instruction only encodes 28 bits of the 32-bit instruction. And it jumps to the absolute address within the same 256MB page as the program counter of the instruction. So the low 28-bits of the address come from the instruction, and the high 4 bits come from the current program counter. [2]

So this case is wrong, and _cannot_ be encoded by the `j` instruction:

```
ipc    = 0x10000000
target = 0x0ffffffc  
```

If your PC is in 256MB `page 1`, you cannot use this instruction to get to `page 0`.

I do believe the in_range value would be computed correctly if we add in the missing parens for operator precedence: 

```
bool in_range = ((ipc ^ static_cast<uint32_t>(target)) > (kImm26Bits + kImmFieldShift)) == 0;
```

Notes:

[1] target valued type of `int32_t` seems a bit wrong to me, but works because MIPS user-mode addresses are in range 0-0x07fff.ffff for mips32.

[2] There is a detail that the PC checked is the PC of the instruction in the delay slot, which is why we add 1 instruction to the pc value: `ipc = reinterpret_cast<uint32_t>(pc_ + 1 \* kInstrSize);

**Comment by ilijapavlovic** [comment]
If the target is a 32-bit (64-bit) pointer (in the specification they could be more specific what is the target) than condition which you wrote has sense. Also, following must be done:

```
bool in_range = ((ipc ^ static_cast<uint32_t>(target)) > (kImm26Bits + kImmFieldShift)) == 0;
GenInstrJump(J, static_cast<uint32_t>(target >> 2) & kImm26Mask);
```

If this is ok, than I have to adapt the test for jump instruction in disassembler.

**Comment by paul99** [comment]
Yes, agreed that the address passed to GenInstrJump() must be masked. Wow, I wonder how that ever worked properly.

Another thing to consider: all the checks are done _only_ in DEBUG mode. I believe that our initial use of J/JAL was only for JALR->JAL optimization, and that check of in-range was done for every use (even in release builds) by the caller. 

If we have callers of J/JAL that do not themselves check (I think the related mozilla fails in https://github.com/paul99/v8m-rb/issues/149 are one example of this) -- then I think these should be release mode checks. What do you think?

**Comment by paul99** [comment]
Also, I think you see it, but you can't just stick some fixed address in j/jal for test-disassembler, you will have to get the address of the PC and construct a legal/valid j/jal instruction, and then generate the expected disassembly as well.

**Comment by ilijapavlovic** [comment]
Yes, I will have to modify test-disassembler and somehow generate/construct j/jal instruction and expected output.

For release mode we expect (and maybe it is required) that callers will provide values inside valid ranges for all instructions. I see that for other instructions such kind of checking was not performed, so why should we make an exception for this case.

**Comment by ilijapavlovic** [comment]
Fixed checking target value. Adapted tests in disassembler.
https://codereview.chromium.org/1237083003

**Comment by ilijapavlovic** [comment]
Landed.
https://chromium.googlesource.com/v8/v8/+/2bc5a2121102282fc67cf96423bae02b19e18e42

**Comment by paul99** [comment]
Ilija - I'm afraid the commit message formatting did not work out correctly. In the future please be very careful with this.

The format should have looked like this:

```
MIPS: Improved checking target ranges for J and JAL instructions.

Adapted disassembler test for J and JAL instructions.

TEST=cctest/test-disasm-mips[64]
BUG=
```

The key point is that the first line of the commit should be a summary description of what changed, possibly prefaced by MIPS: or MIPS64: if those are appropriate [1]. The line should be short, so that it is readable in a one-line lot log, or a git history viewer, or whatever.

Then there must be exactly one blank line.

Then you can include more description, if necessary.

Here is a fairly good reference: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

**Comment by paul99** [comment]
Hi Ilija, it seems we have another problem with j/jal in test-disassembler-mips64, in test Type3. Can you PTAL?

https://v8.mips.com/bbv8/builders/Target%20-%20MIPS64%20-%20edge/builds/1806/steps/Check/logs/Type3

**Comment by ilijapavlovic** [comment]
Adapted disassembler test for J and JAL.
 https://codereview.chromium.org/1258743004

**Comment by ilijapavlovic** [comment]
The 2nd patch where only disassembler is adapted. No changes in disassembler tests.
https://codereview.chromium.org/1258743004

**Comment by paul99** [comment]
This is in CQ, but blocked due to closed tree. Let's verify this on BB once it lands, and then please close it, Ilija.

**Comment by ilijapavlovic** [comment]
Ok.

**Comment by ilijapavlovic** [comment]
Landed.
https://chromium.googlesource.com/v8/v8/+/496bd53bcc9d5d58c5df2e707933b47f2dd2fecd

**Comment by paul99** [comment]
Thanks for the fix! Closing.
